### ISSUE TITLE
Include Babylon plugin names on File AST node?

### ISSUE THREAD

**Issue by loganfsmyth** [babel/babel#7558]
There are a few options here so I'm kind of curious. This seems like the easiest to me though.

I can imagine cases where AST nodes end up being ambiguous, ideally not in final forms, but certainly when multiple proposals are possible like the pipeline proposal. It seems like having at least some information about this in the AST could be helpful.

The other option here would be to actually have a version flag on the AST node, but it seems like that could encourage different parts of the same file to use different forms of an experimental syntax, which seems scary to me. It also means that the `t.foo()` builder functions would need to be versioned somehow.

I think we should basically just do

    file.plugins = parser.plugins;

on the `File` node at https://github.com/babel/babel/blob/master/packages/babylon/src/parser/statement.js#L31.

This would also allow babel-generator to output decorators in the correct location, since right now it is hard-coded for `export @dec class Foo {}`, even if the parser originally parsed `@dec export class Foo {}`, which isn't ideal.

Thoughts?

**Comment by loganfsmyth** [comment]
As a related note from Slack, if we should really include Babylon's major version on `File`, since AST node types for experimental features could change between versions, and while plugins have access to the Babel version, it would still probably make sense to include that metadata on `File`, so that if plugins become part of the standard, things could check if the plugins is in the `file.plugins` data, or that the version is new enough to include it by default. Though really we'd probably want to leave plugins mentioned in `plugins` even if they don't technically exist and have been merged as standard, at least for some semver range to keep things stable.

**Comment by mAAdhaTTah** [comment]
What about "versioning", in some form, the unstable nodes via the plugins? Internal plugins could have their own version numbers, which you could check against in the babel plugin to ensure you're working against the right version of the nodes. So the `pipelineOperator` plugin could be `0.1.0`, and babel plugins can know what the internal nodes look like based on the plugin version, instead of needing to look at something less granular like babylon's version.

**Comment by loganfsmyth** [comment]
Yeah, versioning is kind of a open question for us for Babylon and plugins for experimental stuff. My assumption at the moment is that if we're still in the same major version of Babel/Babylon and want to make a serious incompatible change, we'd need to introduce `pilelineOperator2` or something like that in Babylon. We've been relatively lucky in the past with proposal stability, but it's something we need to start figuring out.

> instead of needing to look at something less granular like babylon's version.

I should clarify that my comment on Babylon version I meant only for major versions, like if someone writes:

```
if (file.plugins["pipelineOperator"]) {}
```
and then we release Babel/Babylon 8.x that removes the plugin entirely because it has been standardized, ideally we want to make it easy for them to do
```
if (file.plugins["pipelineOperator"] && file.version >= 8) {}
```
since it wouldn't be in the `file.plugins` object anymore. If the logic reading the AST lives in a Babel plugin they could check the Babel version, but it's always possible that the AST is being passed off to something else that doesn't know the Babel version that produced the AST.

**Comment by mAAdhaTTah** [comment]
I'm thinking we don't even need to do that. In the plugin code, they can write:

```js
if (semver.satisifies(file.plugins["pipelineOperator"], "^1.0.0")) {}
```

Then we can turn the plugin on by default and as long as we maintain the same version number in `plugins`, no code has to change.

**Comment by loganfsmyth** [comment]
Yep, the

> as long as we maintain the same version number in plugins, no code has to change.

is what I'm on the fence about. It seems like _eventually_ at least we should just remove things? Doesn't have to be in the immediate next version of Babylon or anything though.

Do you have a specific concern with including the Babylon major version on `File`?

> ```
> if (semver.satisifies(file.plugins["pipelineOperator"], "^1.0.0")) {}
> ```

I admit I'm hesitant to expose the full semver range, because if the minor and patch changes are anything that aren't already reflected in the AST structure themselves, doesn't that immediately mean in should have been a major version?

**Comment by mAAdhaTTah** [comment]
> It seems like _eventually_ at least we should just remove things?

My thought was we'd remove merged plugins from the `plugins` list in the next major version, and that this would allow breaking changes in unstable features without needing to bump babylon's major version.

Something like pipeline could undergo some churn as we iron out the details, and versioning its internals could ensure the plugins are running against the correct expected nodes. The way I'm thinking about it, babylon plugins for features in flux aren't part the "official" AST, so those changes get their own version. Once they do land, then the plugin knows it's running against the correct version based on the version in the `package.json`.

> Do you have a specific concern with including the Babylon major version on `File`?

No, that seems fine. I'm wondering how we're thinking about versioning the AST when dealing w/ unstable features.

**Comment by loganfsmyth** [comment]
> My thought was we'd remove merged plugins from the plugins list in the next major version, and that this would allow breaking changes in unstable features without needing to bump babylon's major version.

I think we might be talking past eachother. So I understand, you're saying, in the next major version of Babylon, we'd remove the merged plugins from babylon's plugin list? That I agree with and that's what I'm saying too.

> that this would allow breaking changes in unstable features without needing to bump babylon's major version.

What is "this" in this sentence? I'm confused. If you mean having separate semver versions for the plugin inside Babylon, I agree with that idea, but again I'm not sure that we'd want full minor and patch versions?

> Something like pipeline could undergo some churn as we iron out the details, and versioning its internals could ensure the plugins are running against the correct expected nodes. The way I'm thinking about it, babylon plugins for features in flux aren't part the "official" AST, so those changes get their own version.

Yeah, the main difficulty here is that usually `babel-types` still needs to know _something_ about the structure of the AST in order to analyze the scopes properly, and that logic is all versioned with babel-core directly, so it isn't just Babylon and the transform plugin that need to know about the AST.

Versioning something as frequently as the pipeline operator seems like it might change isn't something we've done before and I'm at least a little worried we're going to get to a place where breaking changes become necessary to Babel.

**Comment by mAAdhaTTah** [comment]
> What is "this" in this sentence? I'm confused. If you mean having separate semver versions for the plugin inside Babylon, I agree with that idea, but again I'm not sure that we'd want full minor and patch versions?

Yeah, this was "internal plugins with their own semver versions". Although we don't necessarily need full semver ranges, as you say, and could just increment a `Number` every time the AST for a given internal plugin changes.

> Versioning something as frequently as the pipeline operator seems like it might change isn't something we've done before and I'm at least a little worried we're going to get to a place where breaking changes become necessary to Babel.

Yeah... I'm trying to help avoid that. I guess I'm suggesting that `babel-types` (and whatever else) would use the `file.plugins` version to know how to analyze something like the pipeline, so the pipeline could have breaking changes, bump the number, and whatever else would know its different, and we remove all of this versioning code only in the next babel version.

This does feel like it's bending / abusing semver, but to conform strictly would be to either put babel in a position of bumping majors every time there's an AST change or denying us the ability to play with the AST for this proposal. Or introducing new flags every time we want to make an AST change :/
